### PR TITLE
Remove EntityProperty, store entity data as KVObject

### PR DESCRIPTION
**Pull request by kristiker**

### CODE REVIEW

File: ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs [hunk 1]
```diff
@@ -207,37 +156,40 @@ private static Entity ParseEntityProperties(byte[] bytes)
...
-                    }
+                    EntityFieldType.Boolean => (KVType.BOOLEAN, (object)dataReader.ReadBoolean()),
+                    EntityFieldType.Float => (KVType.DOUBLE, (double)dataReader.ReadSingle()),
+                    EntityFieldType.Float64 => (KVType.DOUBLE, dataReader.ReadDouble()),
+                    EntityFieldType.Color32 => (KVType.ARRAY, new KVObject("", dataReader.ReadBytes(4).Select(c => new KVValue(KVType.INT64, c)).ToArray())),
+                    EntityFieldType.Integer => (KVType.INT64, dataReader.ReadInt32()),
+                    EntityFieldType.UInt => (KVType.UINT64, dataReader.ReadUInt32()),
+                    EntityFieldType.Integer64 => (KVType.UINT64, dataReader.ReadUInt64()), // TODO: Is supposed to be ReadInt64?
+                    EntityFieldType.Vector or EntityFieldType.QAngle => (KVType.STRING, $"{dataReader.ReadSingle()} {dataReader.ReadSingle()} {dataReader.ReadSingle()}"),
+                    EntityFieldType.CString => (KVType.BOOLEAN, dataReader.ReadNullTermString(Encoding.UTF8)), // null term variable
```
Comment (kristiker) [inline ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs:170]:
```suggestion
                    EntityFieldType.CString => (KVType.STRING, dataReader.ReadNullTermString(Encoding.UTF8)), // null term variable
```

File: ValveResourceFormat/Serialization/KeyValues/KVObject.cs [hunk 2]
```diff
@@ -185,16 +195,43 @@ private static void WriteKey(IndentedTextWriter writer, string key)
...
+        {
+            if (Properties.TryGetValue(name, out var property))
             {
-                return default;
+                var valueObject = property.Value;
+
+                // We typicallly want to get a bool, int, uint, or float property,
+                // however it might be stored as string, which will raise FormatException.
+                // So here we try to convert the string to floating point number.
+                if (typeof(T) != typeof(string) && valueObject is string stringValue)
```
Comment (kristiker) [inline ValveResourceFormat/Serialization/KeyValues/KVObject.cs:217]: Maybe explicit check for bool, int, uint, etc

File: GUI/Types/Renderer/GLWorldViewer.cs [hunk 3]
```diff
@@ -507,10 +507,8 @@ private void ShowEntityProperties(SceneNode sceneNode)
         {
             foreach (var property in sceneNode.EntityData.Properties)
```
Comment (kristiker) [inline GUI/Types/Renderer/GLWorldViewer.cs:508]:
```suggestion
            foreach (var (key, value) in sceneNode.EntityData.Properties)
```

File: GUI/Types/Renderer/WorldLoader.cs [hunk 4]
```diff
@@ -908,47 +817,18 @@ void LoadEntity(string classname, Entity entity)
                 var isAnimated = modelNode.SetAnimationForWorldPreview(animation);
                 if (isAnimated)
                 {
-                    var holdAnimation = entity.GetProperty("holdanimation");
-                    if (holdAnimation != default)
+                    var holdAnimationOn = entity.GetPropertyUnchecked<bool>("holdanimation");
```
Comment (xPaw) [inline GUI/Types/Renderer/WorldLoader.cs:820]: does this actually work for strings of "1"

File: ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs [hunk 5]
```diff
@@ -385,37 +319,22 @@ public string ToForgeGameData()
...
-                        var model = (string)property.Value.Data;
-
                         if (model.Contains("/entities/", StringComparison.Ordinal) || model.Contains("\\entities\\", StringComparison.Ordinal))
                         {
                             brushEntities.Add(classname);
                         }
                     }
 
-                    entityProperties.Add((key, property.Value.Type));
+                    //entityProperties.Add((key, property.Value.Type));
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs:337]: todo?

File: ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs [hunk 6]
```diff
@@ -493,16 +412,16 @@ public string ToForgeGameData()
         }
 
         // TODO: Invert this, and upconvert legacy entity fields into keyvalues
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/EntityLump.cs:414]: done :)

File: GUI/Types/Renderer/WorldLoader.cs (hunk 4 above)
Comment (kristiker) [reply GUI/Types/Renderer/WorldLoader.cs:820]: Yes, `GetPropertyUnchecked` now converts the string to float, which `Convert.ChangeType` is able to convert to boolean.

image

**Pull request merged**